bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -132,18 +133,100 @@ public final class Image extends Resource implements Drawable {
 	 */
 	static final int DEFAULT_SCANLINE_PAD = 4;
 
-	private Map<Integer, ImageHandle> zoomLevelToImageHandle = new HashMap<>();
-
 	private List<Consumer<Image>> onDisposeListeners;
 
+	private final ImageHandleManager imageHandleManager = new ImageHandleManager();
+
+	private class ImageHandleManager {
+		private Map<Integer, ImageHandle> zoomLevelToImageHandle = new ConcurrentHashMap<>();
+		private Set<ImageHandle> imageHandles = new HashSet<>();
+
+		ImageHandle getImageHandle(int zoom) {
+			final ImageHandle imageHandleFuture = zoomLevelToImageHandle.get(zoom);
+			if (imageHandleFuture == null) {
+				return null;
+			}
+			return imageHandleFuture;
+		}
+
+		ImageHandle getHandleOrCreate(int zoom, Supplier<ImageHandle> creator) {
+			ImageHandle imageHandle = zoomLevelToImageHandle.computeIfAbsent(zoom, __ -> {
+				if (zoom == -1) {
+					return null;
+				}
+				ImageHandle imageHandleToReturn = creator.get();
+				return imageHandleToReturn;
+			});
+			return imageHandle;
+		}
+
+		boolean hasImageHandle(ImageHandle imageHandle) {
+		    return zoomLevelToImageHandle.values().contains(imageHandle);
+		}
+
+		boolean hasZoom(int zoom) {
+			return zoomLevelToImageHandle.containsKey(zoom);
+		}
+
+		boolean hasNoBaseHandle() {
+		    return zoomLevelToImageHandle.isEmpty();
+		}
+
+		List<ImageHandle> getAllImageHandles() {
+			return zoomLevelToImageHandle.values().stream().toList();
+		}
+
+		Set<Integer> getAllZooms() {
+			return zoomLevelToImageHandle.keySet();
+		}
+
+		void destroyHandles(Predicate<Integer> filter) {
+			Iterator<Entry<Integer, ImageHandle>> it = zoomLevelToImageHandle.entrySet().iterator();
+			while (it.hasNext()) {
+				Entry<Integer, ImageHandle> zoomToHandle = it.next();
+				if (filter.test(zoomToHandle.getKey())) {
+					ImageHandle imageHandle = zoomToHandle.getValue();
+					it.remove();
+					destroy(imageHandle);
+				}
+			}
+		}
+
+		void destroy(ImageHandle imageHandle) {
+			zoomLevelToImageHandle.remove(imageHandle.zoom, imageHandle);
+			imageHandle.destroy();
+			imageHandles.remove(imageHandle);
+			cleanupDanglingImageHandles();
+		}
```

Comment:
This method feels wrong to me: either the `ImageHandleManager` manages the handles and does not auto destroy (with the other method accepting a Predicate) or a handle is managed separately, but then it should not be artificially be passed over this manager.

If I am not mistaken, at least the `imageHandles` set, the `register` method and the `destroy(ImageHandle)` should be removed.